File: webapp/renderer/site-images.js
/** SerienStream/AniWorld image helpers – mirrors Android SiteImages.kt */
window.SiteImages = (() => {
  function preferJpeg(url) {
    if (!url || !String(url).trim()) return null;
    let u = String(url).trim();
    if (u.startsWith("//")) u = `https:${u}`;
    if (u.startsWith("http://")) u = `https://${u.slice("http://".length)}`;
    if (u.startsWith("data:") || /\.svg(\?|$)/i.test(u)) return null;
    u = upgradeResolution(u);
    if (/\/media\/images\//i.test(u)) {
      u = u.replace(/([?&])format=(avif|webp)/gi, "$1format=jpg");
      if (!/[?&]format=/i.test(u)) {
        u += (u.includes("?") ? "&" : "?") + "format=jpg";
      }
    } else {
      u = u.replace(/\/webp\//gi, "/jpeg/").replace(/\.webp(\?|$)/gi, ".jpg$1");
    }
    return u;
  }

  function upgradeResolution(url) {
    let u = url;
    const kinds = ["channel", "backdrop"];
    for (const kind of kinds) {
      u = u
        .replace(
          new RegExp(`/media/images/${kind}/mobile/`, "gi"),
          `/media/images/${kind}/2x-desktop/`,
        )
        .replace(
          new RegExp(`/media/images/${kind}/tablet/`, "gi"),
          `/media/images/${kind}/2x-desktop/`,
        )
        .replace(
          new RegExp(`/media/images/${kind}/desktop/`, "gi"),
          `/media/images/${kind}/2x-desktop/`,
        )
        .replace(
          new RegExp(`/media/images/${kind}/tile/`, "gi"),
          `/media/images/${kind}/2x-tile/`,
        )
        .replace(
          new RegExp(`/media/images/${kind}/hero-mobile/`, "gi"),
          `/media/images/${kind}/hero-2x-desktop/`,
        )
        .replace(
          new RegExp(`/media/images/${kind}/hero-desktop/`, "gi"),
          `/media/images/${kind}/hero-2x-desktop/`,
        );
    }
    return u;
  }

  function isChannel(url) {
    return /\/media\/images\/channel\//i.test(url || "");
  }
  function isBackdrop(url) {
    return /\/media\/images\/backdrop\//i.test(url || "");
  }

  function resolutionScore(url) {
    if (!url) return -1;
    const u = url.toLowerCase();
    if (u.includes("/2x-desktop/") || u.includes("/2x-tile/") || u.includes("/hero-2x-desktop/"))
      return 40;
    if (u.includes("/orig/")) return 35;
    if (u.includes("/desktop/") || u.includes("/tile/") || u.includes("/hero-desktop/")) return 30;
    if (u.includes("/tablet/")) return 20;
    if (u.includes("/mobile/") || u.includes("/hero-mobile/")) return 10;
    return 5;
  }

  function bestUrl(candidates) {
    return (candidates || [])
      .map((c) => preferJpeg(c))
      .filter(Boolean)
      .sort((a, b) => resolutionScore(b) - resolutionScore(a))[0] || null;
  }

  function pickPoster(...candidates) {
    const list = candidates.flat().filter(Boolean);
    return preferJpeg(list.find(isChannel)) || bestUrl(list) || preferJpeg(list[0]);
  }

  function pickBackdrop(...candidates) {
    const list = candidates.flat().filter(Boolean);
    return preferJpeg(list.find(isBackdrop)) || bestUrl(list) || preferJpeg(list[0]);
  }

  function fromSrcset(srcset) {
    if (!srcset) return null;
    const urls = String(srcset)
      .split(",")
      .map((part) => part.trim().split(/\s+/)[0])
      .filter((u) => u && !u.startsWith("data:"));
    return bestUrl(urls);
  }

  return {
    preferJpeg,
    upgradeResolution,
    pickPoster,
    pickBackdrop,
    fromSrcset,
    isChannel,
    isBackdrop,
    resolutionScore,
    bestUrl,
  };
})();

File: app/src/main/java/com/streamvault/tv/data/catalog/SiteImages.kt
package com.streamvault.tv.data.catalog

/**
 * SerienStream-compatible image helpers.
 * Covers live under `/media/images/channel/...` and `/media/images/backdrop/...`.
 * Prefer JPEG + desktop/2x-desktop for Fire TV (AVIF/WebP and mobile thumbs look pixelated).
 */
object SiteImages {
    fun preferJpeg(url: String?): String? {
        if (url.isNullOrBlank()) return null
        var u = url.trim()
        if (u.startsWith("//")) u = "https:$u"
        u = upgradeResolution(u)
        if (u.contains("/media/images/", ignoreCase = true)) {
            u = u.replace(Regex("""([?&])format=(avif|webp)""", RegexOption.IGNORE_CASE), "$1format=jpg")
            if (!u.contains("format=", ignoreCase = true)) {
                u += if (u.contains("?")) "&format=jpg" else "?format=jpg"
            }
        }
        return u
    }

    /** Prefer 2x-desktop > desktop over mobile/tablet thumbs. */
    fun upgradeResolution(url: String): String {
        var u = url
        val kinds = listOf("channel", "backdrop")
        for (kind in kinds) {
            u = u
                .replace("/media/images/$kind/mobile/", "/media/images/$kind/2x-desktop/", ignoreCase = true)
                .replace("/media/images/$kind/tablet/", "/media/images/$kind/2x-desktop/", ignoreCase = true)
                .replace("/media/images/$kind/desktop/", "/media/images/$kind/2x-desktop/", ignoreCase = true)
        }
        return u
    }

    fun channelCover(baseUrl: String, posterSlug: String): String {
        val base = baseUrl.trimEnd('/')
        val slug = posterSlug.trim().trim('/')
        return preferJpeg("$base/media/images/channel/2x-desktop/$slug")!!
    }

    fun isChannel(url: String?): Boolean =
        url?.contains("/media/images/channel/", ignoreCase = true) == true

    fun isBackdrop(url: String?): Boolean =
        url?.contains("/media/images/backdrop/", ignoreCase = true) == true

    fun resolutionScore(url: String?): Int {
        if (url.isNullOrBlank()) return -1
        val u = url.lowercase()
        return when {
            u.contains("/2x-desktop/") -> 40
            u.contains("/desktop/") -> 30
            u.contains("/tablet/") -> 20
            u.contains("/mobile/") -> 10
            u.contains("/orig/") -> 35
            else -> 5
        }
    }

    fun bestUrl(candidates: Collection<String?>): String? =
        candidates
            .mapNotNull { preferJpeg(it) }
            .maxByOrNull { resolutionScore(it) }

    fun pickPoster(vararg candidates: String?): String? {
        val list = candidates.filterNotNull()
        return preferJpeg(list.firstOrNull { isChannel(it) })
            ?: bestUrl(list)
            ?: preferJpeg(list.firstOrNull { it.isNotBlank() })
    }

    fun pickBackdrop(vararg candidates: String?): String? {
        val list = candidates.filterNotNull()
        return preferJpeg(list.firstOrNull { isBackdrop(it) })
            ?: bestUrl(list)
            ?: preferJpeg(list.firstOrNull { it.isNotBlank() })
    }

    /** From a srcset attribute, pick the highest-res candidate. */
    fun fromSrcset(srcset: String?): String? {
        if (srcset.isNullOrBlank()) return null
        val urls = srcset.split(',')
            .mapNotNull { part ->
                part.trim().substringBefore(' ').trim().takeIf { it.isNotBlank() && !it.startsWith("data:") }
            }
        return bestUrl(urls)
    }
}
